pkg/runners/zero/zero.go

```diff
@@ -240,6 +268,98 @@ func (runner *ZeroRunner) RunFor(steps uint64) error {
 	return nil
 }
 
+// EndRun is responsible for running the additional steps after the program was executed,
+// until the checkUsedCells doesn't return any error.
+// Since this vm always finishes the run of the program at the number of steps that is a power of two in the proof mode,
+// there is no need to run additional steps before the loop.
+func (runner *ZeroRunner) EndRun() {
+	if runner.proofmode {
+		for runner.checkUsedCells() != nil {
+			pow2Steps := utils.NextPowerOfTwo(runner.vm.Step + 1)
+			if err := runner.RunFor(pow2Steps); err != nil {
+				panic(err)
+			}
+		}
+	}
+}
+
+// checkUsedCells returns error if not enough steps were made to allocate required number of cells for builtins
+// or there are not enough trace cells to fill the entire range check range
+func (runner *ZeroRunner) checkUsedCells() error {
+	for _, bRunner := range runner.Layout.Builtins {
+		builtinSegment, ok := runner.vm.Memory.FindSegmentWithBuiltin(bRunner.Runner.String())
+		if ok {
+			_, err := bRunner.Runner.GetAllocatedSize(builtinSegment.Len(), runner.steps())
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return runner.checkRangeCheckUsage()
+}
+
+func (runner *ZeroRunner) checkRangeCheckUsage() error {
+	rcMin, rcMax := runner.getPermRangeCheckLimits()
+	var rcUnitsUsedByBuiltins uint64
+	for _, builtin := range runner.program.Builtins {
+		bRunner := builtins.Runner(builtin)
+		rangeCheckRunner, ok := bRunner.(*builtins.RangeCheck)
+		if ok {
+			rangeCheckSegment, ok := runner.vm.Memory.FindSegmentWithBuiltin(rangeCheckRunner.String())
+			if ok {
+				rcUnitsUsedByBuiltins += rangeCheckSegment.Len() * rangeCheckRunner.RangeCheckNParts
+			}
+		}
+	}
+	unusedRcUnits := (runner.Layout.RcUnits-3)*runner.vm.Step - rcUnitsUsedByBuiltins
+	rcUsageUpperBound := rcMax - rcMin
+	if unusedRcUnits < rcUsageUpperBound {
+		return fmt.Errorf("RangeCheck usage is %d, but the upper bound is %d", unusedRcUnits, rcUsageUpperBound)
+	}
+	return nil
+}
+
+func (runner *ZeroRunner) getPermRangeCheckLimits() (uint64, uint64) {
+	rcMin, rcMax := runner.vm.RcLimitsMin, runner.vm.RcLimitsMax
+
+	for _, builtin := range runner.program.Builtins {
+		bRunner := builtins.Runner(builtin)
+		rangeCheckRunner, ok := bRunner.(*builtins.RangeCheck)
+		if ok {
+			rangeCheckSegment, ok := runner.vm.Memory.FindSegmentWithBuiltin(rangeCheckRunner.String())
+			if ok {
+				rangeCheckUsageMin, rangeCheckUsageMax := rangeCheckRunner.GetRangeCheckUsage(rangeCheckSegment)
+				if rangeCheckUsageMin < rcMin {
+					rcMin = rangeCheckUsageMin
+				}
+				if rangeCheckUsageMax > rcMax {
+					rcMax = rangeCheckUsageMax
+				}
+			}
```

Comment:
I see that at the end you are not only comparing the felts in 16 bits pieces, but also the instructions. Not sure I understand why this is necessary. Could you add some comments explaining?